scripts/zkai_attention_kv_stwo_native_two_head_bounded_weighted_proof_input.py

```
@@ -0,0 +1,666 @@
+            require_kv_item(candidate, label=f"step[{global_step_index}].candidate[{candidate_index}]")
+            for candidate_index, candidate in enumerate(head_candidates)
+            if candidate["position"] <= step["token_position"]
+        ]
+        if not allowed:
+            raise AttentionKvTwoHeadBoundedWeightedInputError(f"input_steps[{global_step_index}] has no allowed candidates")
+        scored = [(candidate, dot(step["query"], candidate["key"])) for candidate in allowed]
+        selected_score = max(score for _, score in scored)
+        weights = [weight_from_gap(selected_score - score) for _, score in scored]
+        denominator = sum(weights)
+        if denominator <= 0:
+            raise AttentionKvTwoHeadBoundedWeightedInputError(f"input_steps[{global_step_index}] has invalid weight denominator")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)